providers/adfs.go

```diff
@@ -65,36 +72,48 @@ func (p *ADFSProvider) GetLoginURL(redirectURI, state, nonce string) string {
 		extraParams.Add("nonce", nonce)
 	}
 	loginURL := makeLoginURL(p.Data(), redirectURI, url.QueryEscape(state), extraParams)
-	if p.SkipScope {
+	if p.skipScope {
 		q := loginURL.Query()
 		q.Del("scope")
 		loginURL.RawQuery = q.Encode()
 	}
 	return loginURL.String()
 }
 
-// EnrichSession to add email
+// EnrichSession calls the OIDC ProfileURL to backfill any fields missing
+// from the claims. If Email is missing, falls back to ADFS `upn` claim.
 func (p *ADFSProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
-	if s.Email != "" {
-		return nil
+	err := p.oidcEnrichFunc(ctx, s)
+	if err != nil || s.Email == "" {
+		// OIDC only errors if email is missing
+		return p.fallbackUPN(ctx, s)
+	}
+	return nil
+}
+
+// RefreshSession refreshes via the OIDC implementation. If email is missing,
+// falls back to ADFS `upn` claim.
```

Comment:
Fall backs like this make the user side behaviour unpredictable. This means I could auth as two different users and get two different sources, which is not ideal.

Is this existing behaviour? We should make an issue to get rid of this

Reply:
This provider hasn't hit a release yet. The current implementation to override missing emails with `upn` changes the provider-wide config based on a user session (a definite bug).

You bring up a very good point though 🤔 

2 options I think:
- We don't let refreshes change the email (either email vs upn from first login is static).
- We remove this conditional `upn` fallback and require users to setup the `oidc-email-claim` if they want to use `upn`.